### PR TITLE
dataflow/gpu: add dataflow prime gpu hints

### PR DESCRIPTION
**Pull request by davidcavazos**
## Description

Use Dataflow Prime to give hints for GPU usage.

R: @tvalentyn
R: @rezarokni

## Checklist
- [ ] I have followed [Sample Guidelines from AUTHORING_GUIDE.MD](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md)
- [ ] README is updated to include [all relevant information](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#readme-file)
- [ ] **Tests** pass:   `nox -s py-3.6` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#test-environment-setup))
- [ ] **Lint** pass:   `nox -s lint` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#test-environment-setup))
- [ ] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [ ] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [ ] Please **merge** this PR for me once it is approved.
- [ ] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/.github/CODEOWNERS) with the codeowners for this sample

### CODE REVIEW

File: dataflow/gpu-examples/tensorflow-landsat/Dockerfile
Comment (davidcavazos) [inline dataflow/gpu-examples/tensorflow-landsat/Dockerfile:19]: As per https://www.tensorflow.org/install/source#gpu for TensorFlow 2.7 we **must** use CUDA 11.2, not the latest. Otherwise it fails.

File: dataflow/gpu-examples/tensorflow-landsat/Dockerfile
Comment (davidcavazos) [inline dataflow/gpu-examples/tensorflow-landsat/Dockerfile:43]: We might want to wait until this fix is published before we merge, probably by Beam 2.35 (?)

Comment (davidcavazos) [reply dataflow/gpu-examples/tensorflow-landsat/Dockerfile]: This works with Beam 2.35 RC3

File: dataflow/gpu-examples/tensorflow-landsat/main.py
Comment (tvalentyn) [inline dataflow/gpu-examples/tensorflow-landsat/main.py:332]: technically, we have one resource hint (--accelerator), that captures both type and count. you could still customize type and count separately if you prefer, but perhaps don't call it a 'hint' to avoid confusion.

**Review by tvalentyn — COMMENTED** [review]
Main comment: looks like we replace the non-prime example with an example that uses Dataflow Prime.

It may be better to have both examples. In which case, the non-prime example can be a part of GPU tutorial, and Prime example could be something that additionally demonstrates some benefits of Dataflow Prime and a part of Prime tutorial.

File: dataflow/gpu-examples/tensorflow-landsat/Dockerfile
Comment (davidcavazos) [reply dataflow/gpu-examples/tensorflow-landsat/Dockerfile:40]: Good point! I'll add a note in the README

File: dataflow/gpu-examples/tensorflow-landsat/main.py
Comment (davidcavazos) [reply dataflow/gpu-examples/tensorflow-landsat/main.py:332]: I would prefer to capture them separately to avoid user errors as much as possible. That way the accelerator hint format can stay as an implementation detail. I'll remove the word "hint".

File: dataflow/gpu-examples/tensorflow-landsat/main.py
Comment (rezarokni) [inline dataflow/gpu-examples/tensorflow-landsat/main.py:278]: Explain with comments why we are using Reshuffle, can link the break fusion page in Dataflow docs

File: dataflow/gpu-examples/tensorflow-landsat/main.py
Comment (rezarokni) [inline dataflow/gpu-examples/tensorflow-landsat/main.py:69]: Link to docs which show different types.

File: dataflow/gpu-examples/tensorflow-landsat/run.yaml
Comment (rezarokni) [inline dataflow/gpu-examples/tensorflow-landsat/run.yaml:48]: As this needed a 13 Gig to 1 mem ratio before, maybe worth adding a memory min limit to the hint as well

Comment (davidcavazos) [reply dataflow/gpu-examples/tensorflow-landsat/run.yaml:48]: Yes, I actually added a `--resource_hints=min_ram=32GB`. I tried 16GB but I was still hitting out of memory errors. Something changed where it tried to load more images per worker I guess.

**Comment by davidcavazos** [comment]
@tvalentyn I also thought of having two separate examples, but it meant too much duplicated code. That would be maintaining two almost identical samples. Maybe what we can do is having a config file to run with Prime and a config file to run without it. With that in mind, we would have to pass the accelerator flag directly as you suggested.

But that would involve having the accelerator flag passed to the entire pipeline instead of having the per-transform hints. Or maybe we can have the per-transform hints, but pass None if it's in non-Prime. What do you think?

File: dataflow/gpu-examples/tensorflow-landsat/run.yaml
Comment (tvalentyn) [reply dataflow/gpu-examples/tensorflow-landsat/run.yaml:48]: I may be running 2 cores on the VM? I think we may need to set --experiments=no_use_multiple_sdk_containers

**Comment by tvalentyn** [comment]
> @tvalentyn I also thought of having two separate examples, but it meant too much duplicated code. That would be maintaining two almost identical samples. Maybe what we can do is having a config file to run with Prime and a config file to run without it.

We currently have two GPU offerings: GPUs without prime and GPUs with prime, it's different use-cases in Dataflow. Prime is in Preview, while Dataflow GPUs are GA. I think right now we cannot recommend prime as a drop in replacement for non-prime, so it may be good to show users how to use both.

We could decide that a non-prime GPU offering does not need a tutorial and the documentation is sufficient. then we can move GPU tutorial to use Prime and have only one example. 

Alternatively, we can have two tutorials. they may not necessarily be the same. Ideally, the GPU prime tutorial would focus on prime advantages, for example demonstrate that having multiple worker pools results in cost savings, while running the same job on a single pool is more expensive. 

would leave that to DPE folks to decide.

File: dataflow/gpu-examples/tensorflow-landsat/run.yaml
Comment (davidcavazos) [reply dataflow/gpu-examples/tensorflow-landsat/run.yaml:48]: It's already using it in line 53

**Comment by davidcavazos** [comment]
@tvalentyn I like the "We could decide that a non-prime GPU offering does not need a tutorial and the documentation is sufficient. then we can move GPU tutorial to use Prime and have only one example" alternative.

We also have the `tensorflow-minimal` and `pytorch-minimal` samples, which are non-prime. So how about we use the Landsat tutorial as the prime example, and simply link to the minimal examples for the non-prime examples for now.

We still need to update the Landsat tutorial in the docs for the prime setup and instructions.

How would you feel about this?

**Comment by tvalentyn** [comment]
Once prime is GA that is completely fine; while prime is preview, but GPU is GA , it's still possible, but we may need extra wording around this in the tutorial

**Comment by davidcavazos** [comment]
@pcoet what are your thoughts?

File: dataflow/gpu-examples/tensorflow-landsat/main.py
Comment (davidcavazos) [reply dataflow/gpu-examples/tensorflow-landsat/main.py:278]: Thanks, adding a comment and linking to https://cloud.google.com/dataflow/docs/guides/deploying-a-pipeline#preventing-fusion

File: dataflow/gpu-examples/tensorflow-landsat/main.py
Comment (davidcavazos) [reply dataflow/gpu-examples/tensorflow-landsat/main.py:69]: Thanks, I added a comment and a link to https://cloud.google.com/dataflow/docs/resources/locations#gpu_availability

I also added a similar note in the README when running the pipeline and passing the GPU type.

**Comment by pcoet** [comment]
@davidcavazos I don't have a lot of context here, but my initial thought is that we should keep both the Prime and non-Prime examples for now. That seems like the best solution for users.

File: dataflow/gpu-examples/tensorflow-landsat/Dockerfile
Comment (leahecole) [reply dataflow/gpu-examples/tensorflow-landsat/Dockerfile:19]: Can you please add this as an inline comment so that other maintainers and I don't upgrade it by accident?

**Comment by davidcavazos** [comment]
Alright, sounds good to me. I'll create a copy and we'll take it from there.

**Comment by davidcavazos** [comment]
The sample was moved to `tensorflow-landsat-prime` and left the original `tensorflow-landsat` sample untouched. After GA we might want to revisit if we want to adapt the tutorial in the docs, point to the prime sample, and remove the old sample.

**Comment by davidcavazos** [comment]
Tests are passing, but we should still wait until 2.35 stable is published before merging.

**Comment by rezarokni** [comment]
+1 on two copies, but fine if you use same code with branches.

**Comment by davidcavazos** [comment]
Hi @leahecole, Beam 2.35 stable is out, tests are passing with it. I think this should be ready to merge.

**Review by kurtisvg — APPROVED** [review]

File: dataflow/gpu-examples/tensorflow-landsat/Dockerfile
Comment (kurtisvg) [reply dataflow/gpu-examples/tensorflow-landsat/Dockerfile]: Should we consider pinning this in renovate?

**Review by leahecole — APPROVED** [review]

**Pull request merged**
